### PR TITLE
CLOUDP-82159: Upgrade to operator-sdk v1.4.0

### PR DESCRIPTION
**Pull request by alyacb**
This change re-scaffolds the operator using the latest version of operator-sdk (v1.4.0)
- Code refactorings:
   - pkg/apis/mongodb/v1/ -> api/v1/
   - pkg/controller/ -> controllers/
- CRD refactorings:
   - CRDs were moved to config/, CRs to config/samples
   - The new directory structure has roles in config/rbac/
   - operator.yaml was renamed and placed in config/manager/manager.yaml
   - kustomize is being used everywhere
- Build refactorings:
   - Added a Makefile. The operator can now be built and deployed locally using:
     `make generate manifests docker-build docker-push deploy IMG=localhost:5000/mongodb-kubernetes-operator`
      - `make manifests` generates CRDs
      - `make generate` generates deepcopy code
      - `make docker-build docker-push` builds the docker image and pushes it. I modified this directive to generate the Dockerfile using our existing scripts, but if you want to generate a Dockerfile other than the operator, e.g. e2e, you can now run `make docker-build docker-push IMG=localhost:5000/e2e DOCKERFILE=e2e`
   - Changes to the operator Dockerfile template to build the code in the same way as the Makefile
   - Changes to any affected scripts so they can work with the new directory format/ way of building

It was necessary to remove the wrappers around the StatefulSetConfiguration field in mongodb_types, as fields annotated with `"-"` json struct tags no longer are included, not even with the custom JSON marshalling code. This reintroduced CRD pollution (and accounts for most of the liens of code changed). The CRD is automatically generated, so hopefully this shouldn't be too much of a problem.
```golang
StatefulSetConfiguration appsv1.StatefulSetSpec `json:"statefulSet,omitempty"`
```

For a similar reason, these changes break the additionalMongodConfig field(the only failing e2e test is `e2e_test_replica_set_mongod_config`). However, the fix for this field is not apparent, since its type boils down to `map[string]interface{}`, which cannot be generated in the CRDs as `interface{}` is not a named type.

*Note:* this doesn't include the necessary documentation changes, which should be added in a separate PR.

### CODE REVIEW

File: .evergreen.yml
Comment (alyacb) [inline .evergreen.yml:22]: No longer needed, so I removed this step.

File: api/v1/mongodbcommunity_types.go
Comment (alyacb) [inline api/v1/mongodbcommunity_types.go:72]: This change fixed the statefulset_arbitrary e2e tests. For some reason, even though the custom marshal/unmarshal  code should be including this field when annotated with "-", it is being ignored. For that reason I reintroduced the CRD generation (which should be fine as long as we can automatically generate the CRD).

File: build/bin/entrypoint
Comment (alyacb) [inline build/bin/entrypoint:13]: exec was not working, so I switched to "./"

File: Makefile
Comment (chatton) [inline Makefile:77]: is this being used?

File: scripts/dev/templates/Dockerfile.e2e
Comment (chatton) [inline scripts/dev/templates/Dockerfile.e2e:5]: I think we can actually just remove the CMD block entirely

File: scripts/dev/templates/Dockerfile.operator
Comment (chatton) [inline scripts/dev/templates/Dockerfile.operator:11]: I think this should just belong in the e2e image / unit tests

File: scripts/dev/templates/Dockerfile.operator
Comment (chatton) [inline scripts/dev/templates/Dockerfile.operator:33]: we shouldn't need this

File: Makefile
Comment (chatton) [inline Makefile:53]: [q] maybe we can change this to accept a single test to run? We won't usually want to run more than one test at once

File: config/crd/kustomizeconfig.yaml
Comment (chatton) [inline config/crd/kustomizeconfig.yaml:1]: we're unsure of what this is for - leave it for now.

File: config/samples/arbitrary_statefulset_configuration/mongodb.com_v1_custom_volume_cr.yaml
Comment (chatton) [inline config/samples/arbitrary_statefulset_configuration/mongodb.com_v1_custom_volume_cr.yaml:25]: we need to keep it as statefulSet.spec

File: config/samples/kustomization.yaml
Comment (chatton) [inline config/samples/kustomization.yaml:1]: [q] can we delete this?

File: config/samples/mongodbcommunity_v1_mongodbcommunity.yaml
Comment (chatton) [inline config/samples/mongodbcommunity_v1_mongodbcommunity.yaml:1]: [q] can we delete?

File: api/v1/mongodbcommunity_types.go
Comment (chatton) [reply api/v1/mongodbcommunity_types.go:72]: we need to make sure the end resource still works by specifying `statefulSet.spec`

File: controllers/replica_set_controller.go
Comment (chatton) [inline controllers/replica_set_controller.go:120]: [q] is there a reason this is a method and not a function? Does this implement some interface that is required?

Not a big deal either way!

**Review by chatton — COMMENTED** [review]
Looks great, massive work! Had a few comments, the main issue we need to address is fixing the test failures in the additionalMongodConfig test and making sure that the spec of the MongoDBCommunity resources is still `mdb.spec.statefulSet.spec.*`

File: test/e2e/e2eutil.go
Comment (alyacb) [reply test/e2e/e2eutil.go:195]: Will be removed with switch back to v1beta1.

File: test/e2e/statefulset_arbitrary_config/statefulset_arbitrary_config_test.go
Comment (alyacb) [reply test/e2e/statefulset_arbitrary_config/statefulset_arbitrary_config_test.go:37]: Will be rveerted to the old form with v1beta1 change.

File: controllers/replica_set_controller.go
Comment (alyacb) [reply controllers/replica_set_controller.go:120]: I don't think it's required, but it is the way the scaffolding intended (and matches the Reconcile method). I don't think this particular method gets called anywhere outside of main.go (e.g. by the library), but my vote would be to keep it consistent with operator-sdk's expectations so that we don't diverge too far, just in case they introduce code expecting it to follow this format.

File: test/e2e/e2eutil.go
Comment (alyacb) [reply test/e2e/e2eutil.go]: Done!

File: test/e2e/statefulset_arbitrary_config/statefulset_arbitrary_config_test.go
Comment (alyacb) [reply test/e2e/statefulset_arbitrary_config/statefulset_arbitrary_config_test.go]: Done!

File: .gitignore
Comment (alyacb) [reply .gitignore]: Done!

File: Makefile
Comment (alyacb) [reply Makefile]: No! Good catch, and it is gone now. It would be nice to have a directive just for evergreen someday (when it can be made to get along with Makefiles/ newer go versions perhaps?)

File: scripts/dev/templates/Dockerfile.e2e
Comment (alyacb) [reply scripts/dev/templates/Dockerfile.e2e]: Done!

File: scripts/dev/templates/Dockerfile.operator
Comment (alyacb) [reply scripts/dev/templates/Dockerfile.operator]: You're right! Its gone.

File: Makefile
Comment (alyacb) [reply Makefile]: Good thinking! You can now pass which tests you want to run via the TEST variable, e.g.:
```shell
make test TEST=./controllers/...
```
Note that it still defaults to all tests, so you can still run `make test` as before.

File: config/samples/arbitrary_statefulset_configuration/mongodb.com_v1_custom_volume_cr.yaml
Comment (alyacb) [reply config/samples/arbitrary_statefulset_configuration/mongodb.com_v1_custom_volume_cr.yaml]: Done!

Comment (alyacb) [reply config/samples/arbitrary_statefulset_configuration/mongodb.com_v1_custom_volume_cr.yaml]: Done!

File: config/samples/kustomization.yaml
Comment (alyacb) [reply config/samples/kustomization.yaml]: Yes, and done!

File: config/samples/mongodbcommunity_v1_mongodbcommunity.yaml
Comment (alyacb) [reply config/samples/mongodbcommunity_v1_mongodbcommunity.yaml]: Yes! Deleted.

**Comment by alyacb** [comment]
Pushed changes to move back to CRD version v1beta1 and address feedback. Note that whenever we are ready to move to version v1beta1, here's what needs to happen:
1. Replace all occurrences of v1beta1 in the code with v1
2. Change this line in the Makefile from:
```make
CRD_OPTIONS ?= "crd:trivialVersions=true,preserveUnknownFields=true,crdVersions=v1beta1"
```
 to:
```make
CRD_OPTIONS ?= "crd:trivialVersions=true,preserveUnknownFields=false"
```
3. Add the following annotation to the `StatefulSetConfiguration` and `AdditionalMongodConfig` fields on the MongoDBCommunitySpec struct. This is the v1 replacement for `preserveUnknownFields`.
```golang
// +kubebuilder:validation:XPreserveUnknownFields
```
4. Run `make generate manifests`

File: config/manager/kustomization.yaml
Comment (chatton) [inline config/manager/kustomization.yaml:11]: it looks like the `newName` here will be what is substituted in the resulting deployment, we should make sure that we update our release steps with this extra file we need to change (until the whole thing is automated!)

Comment (alyacb) [reply config/manager/kustomization.yaml:11]: This will need a fairly large number of documentation changes around build/deploy process as well. I think maybe this should be part of a separate PR, if only to avoid making this PR any bigger 😅

**Review by chatton — APPROVED** [review]
This looks amazing! I think we're good to go, I think if @bznein / @irajdeep  can take a look at this as well that would be great.

We can tackle the upgrade from v1beta1 to v1 in a future PR.

cc @rodrigovalin this will be a great PR to use as a reference when working on the CRD project.

EDIT: please address the comment regarding the release process before merging!

File: config/manager/kustomization.yaml
Comment (chatton) [reply config/manager/kustomization.yaml:11]: Ah okay, if you could create a ticket we can do that seperately.

Comment (alyacb) [reply config/manager/kustomization.yaml:11]: Done!

File: cmd/manager/main.go
Comment (irajdeep) [inline cmd/manager/main.go:8]: I don't think there is a need to add `controllers` explicitly since it will be used from the import path

Comment (irajdeep) [inline cmd/manager/main.go:8]: I don't think there is a need to add `controllers` explicitly since it will be used from the import path

File: cmd/manager/main.go
Comment (irajdeep) [inline cmd/manager/main.go:15]: same as above.

File: config/crd/kustomization.yaml
Comment (irajdeep) [inline config/crd/kustomization.yaml:2]: q: are we adding `kustomize` support with this PR as well?

File: controllers/replica_set_controller.go
Comment (irajdeep) [inline controllers/replica_set_controller.go:150]: nit: although the context is not used here can call still name this variable `ctx` to be explicit.

**Review by irajdeep — COMMENTED** [review]
LGTM! Great work. Added some minor comments.

File: config/crd/kustomization.yaml
Comment (alyacb) [reply config/crd/kustomization.yaml:2]: Yes. The project was re-scaffolded with the new version of operator-sdk, which generated a bunch of kustomize files for us. The make deploy target in the Makefile uses kustomize to generate the yaml it applies.

File: controllers/replica_set_controller.go
Comment (alyacb) [reply controllers/replica_set_controller.go]: Done!

File: cmd/manager/main.go
Comment (alyacb) [reply cmd/manager/main.go]: Done!

File: cmd/manager/main.go
Comment (alyacb) [reply cmd/manager/main.go]: Done!

**Review by irajdeep — APPROVED** [review]

**Pull request merged**
